README.md
tfs
=======

![logo](https://raw.github.com/fengmk2/tfs/master/logo.png)

[TFS: Taobao FileSystem](http://code.taobao.org/p/tfs/src/) nodejs client.

## Install

```bash
$ npm install tfs
```

## Usage

```js
var tfs = require('tfs');

// var client = tfs.createClient({
//   rootServer: '$host:port',
//   appkey: '$your_appkey',
//   imageServers: ['$your_server1', '$your_server2']
// });

// e.g.: test environment on tb daily
var client = tfs.createClient({
  rootServer: '$tfsdomain:$port',
  appkey: 'appkey',
  imageServers: [
    'img01.tfs.com',
    'img02.tfs.com'
  ]
});

// upload normal file
client.upload(filepath, function (err, info) {
  console.log(info);
  // {
  //   name: 'T2xRETBgZv1RCvBVdK.jpg',
  //   url: 'http://img1.tfs.com/tfscom/T2xRETBgZv1RCvBVdK.jpg',
  //   size: 1024
  // }
});

// upload file content buffer directly
client.upload(fs.readFileSync(filepath), '', function (err, info) {
  console.log(info);
  // {
  //   name: 'T1Pj9zXkRYXXXXXXXX',
  //   url: 'http://img1.tfs.com/tfscom/T1Pj9zXkRYXXXXXXXX',
  //   size: 5929
  // }
});

// upload a stream
client.upload(fs.createReadStream(filepath), 'jpg', function (err, info) {
  console.log(info);
  // {
  //   name: 'T2xRETBgZv1RCvBVdL.jpg',
  //   url: 'http://img1.tfs.com/tfscom/T2xRETBgZv1RCvBVdL.jpg',
  //   size: 1024
  // }
});

// upload custom name file
client.uploadFile(filepath, '320', 'foo.jpg', function (err, info) {
  console.log(info);
  // {
  //   name: 'L1/1/320/foo.jpg',
  //   url: 'http://img1.tfs.com/L1/1/320/foo.jpg',
  //   size: 1984
  // }
});

// upload private file, will not get by CDN
client.uploadPrivate(filepath, function (err, info) {
  console.log(info);
  // {
  //   name: 'T1XeRfXkBXXXXXXXXX.tfsprivate.jpg',
  //   size: 11779,
  //   url: 'http://img01.daily.taobaocdn.net/tfscom/T1XeRfXkBXXXXXXXXX.tfsprivate.jpg'
  // }
});
```

## API v1: 原生TFS

```js
/**
 * Create TFS RESTFul client.
 *
 * @param {Object} options
 *  - {String} appkey
 *  - {String} [appLocation], default is 'tfscom'.
 *  - {Number} [uploadTimeout], upload max timeout, default is 60s.
 *  - {String} [rootServer], 'host:port' format, default is 'restful-store.vip.tbsite.net:3800'.
 *  - {Array} [imageServers], default is CDN online servers list.
 * @return {Client}
 */
function createClient(options);

/**
 * Upload a file.
 * @param {String|Buffer} filepath file path or file content buffer.
 * @param {String} extname. optional, if filepath is buffer or stream type, use this extname.
 * @param {Function(err, info)} callback
 *  - {Object} info
 *   - {String} name, tfs file name
 *   - {String} url, CDN url
 *   - {Number} size, file size
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.upload = function (filepath, extname, callback, timeout);

/**
 * Remove a file by name.
 *
 * @param {String} name
 * @param {Object} [options]
 *  - {Number|String} hide, 1: hidden, 0: show
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.remove = function (name, options, callback, timeout);

/**
 * Download a file by name.
 *
 * @param {String} name
 * @param {String|WriteStream} savefile, save file path or writable stream.
 * @param {Object} [options]
 *  - {Number} offset
 *  - {Number} size
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.download = function (name, savefile, options, callback, timeout);

/**
 * Get file meta.
 * @param {String} name
 * @param {Object} [options]
 *  - {Number} type, 0: normal mode, 1: force mode, get meta info even it was deleted.
 * @param {Function(err, meta)} callback
 * @param {Number} timeout
 */
Client.prototype.getMeta = function (name, options, callback, timeout);
```

## API v2: 自定义文件名

```js
/**
 * Get custom file meta.
 *
 * @param {String} uid user id
 * @param {String} filename custom file name
 * @param {Function(err, meta)} callback
 * @param {Number} timeout
 */
Client.prototype.getFileMeta = function (uid, filename, callback, timeout);

/**
 * Create a file.
 * @param {String} uid, user id
 * @param {[type]} filename, e.g.: 'foo/bar.jpg', 'foo.png'.
 * @param {Function(err, success)} callback
 * @param {Number} [timeout], request timeout
 */
Client.prototype.createFile = function (uid, filename, callback, timeout);

/**
 * Upload a file with custom filename.
 *
 * @param {String} filepath
 * @param {String} uid, user id
 * @param {String} filename
 * @param {Object} [options]
 *  - {Number} offset
 * @param {Function(err, info)} callback
 * @param {Number} [timeout]
 */
Client.prototype.uploadFile = function (filepath, uid, filename, options, callback, timeout);

/**
 * Download a file by custom name.
 *
 * @param {String} uid
 * @param {String} filename
 * @param {String|WriteStream} savefile, save file path or writable stream.
 * @param {Object} [options]
 *  - {Number} offset
 *  - {Number} size
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.downloadFile = function (uid, filename, savefile, options, callback, timeout);

/**
 * Remove a custom name file.
 *
 * @param {String} uid
 * @param {String} filename
 * @param {Function(err, success)} callback
 * @param {Number} [timeout]
 */
Client.prototype.removeFile = function (uid, filename, callback, timeout);
```

## Events

* `ready`: If client init done, will `client.emit('ready')`.
* `error`: If client init error, will `client.emit('error', err)`. User must listen this event once, otherise node process will throw it.
* `refreshError`: If `refreshServers()` error, will `client.emit('refreshError', err)`.
* `servers`: If got new servers list, will `client.emit('servers', servers)`.

## Authors

```bash
$ git summary

 project  : tfs
 repo age : 1 year, 1 month
 active   : 21 days
 commits  : 60
 files    : 18
 authors  :
    34  苏千                  56.7%
    22  fengmk2                 36.7%
     2  dead_horse              3.3%
     2  不四                  3.3%
```

## License

(The MIT License)

Copyright (c) 2012 - 2013 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.2.3 / 2014-02-12 
==================

  * update readme
  * add download private file test cases

0.2.2 / 2014-01-08 
==================

  * fix test cases fail on remove()
  * add uploadPrivate() to support private file (@dead-horse)
  * no default root server
  * update config
  * improve coverage

0.2.0 / 2013-12-20 
==================

  * add v2 api: downloadFile(), getFileMeta()
  * fix test cases
  * install from cnpm
  * fix test cases
  * add toast
  * fixed test cases; change bt to buffer-type

0.1.6 / 2013-08-13 
==================

  * try to guess content-type and extention when extname missing
  * upgrade urllib version

0.1.5 / 2013-08-12 
==================

  * fixed upload buffer directly bug
  * add test usage for test env
  * add diagnose tools link

0.1.4 / 2013-07-11 
==================

  * update daily host; fixed deps version
  * fixed #3 update readme
  * update readme events

0.1.3 / 2013-06-03 
==================

  * refresh servers every 10 minutes. fixed #2
  * use blanket instead of jscover

0.1.2 / 2013-03-11 
==================

  * only support node@0.8.0+

0.1.1 / 2013-03-11 
==================

  * Merge pull request #1 from dead-horse/stream
  * support stream and buffer upload

0.1.0 / 2012-12-14 
==================

  * support getMeta()
  * fixed download test cases
  * support download() file
  * add MIT License
  * publish to github

0.0.4 / 2012-12-01 
==================

  * support remove() file by name

0.0.3 / 2012-11-28 
==================

  * add refreshError listen
  * add demo
  * rename tfs.js to client.js
  * support removeFile
  * coverage getAppid()
  * add uploadFile()
  * update readme

0.0.2 / 2012-11-28 
==================

  * rename appname to appLocation
  * change report to tap
  * fixed repository url

0.0.1 / 2012-11-28 
==================

  * upload(filepath, callback)
  * 完善单元测试
  * init


demo/hello.js
/*!
 * tfs - demo/hello.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var path = require('path');
var tfs = require('../');

var client = tfs.createClient({
  rootServer: '$domain:$port',
  appkey: 'appkey',
  imageServers: [
    'img01.daily.taobaocdn.net',
    'img02.daily.taobaocdn.net'
  ]
});

client.on('servers', function (servers) {
  console.log(servers);
});

client.on('refreshError', function (err) {
  throw err;
});

var logopath = path.join(path.dirname(__dirname), 'logo.png');

// client.uploadFile(logopath, '320', 'tfs.png', function (err, info) {
//   console.log('uploadFile info: %j', info);
//   client.getMeta(info.name, function (err, meta) {
//     console.log(err)
//     console.log('uploadFile meta: %j', meta);
//   });
// });

client.upload(logopath, function (err, info) {
  console.log('upload info: %j', info);
  client.getMeta(info.name, function (err, meta) {
    // console.log(err)
    console.log('upload meta: %j', meta);
  });
});


lib/command.js
/*!
 * tfs - lib/command.js
 * Copyright(c) 2012 - 2013 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var path = require('path');
var util = require('util');

function Command(client, args) {
  this._args = args || {};
  this._client = client;
  this.ignoreStatus = null;
}

Command.prototype._getURL = function () {
  var url = '/v2/' + this._client.appkey + '/' + this._client.appid + '/' +
    this._args.uid + '/file/' + this._args.filename;
  return url;
};

Command.prototype.__defineGetter__('url', function () {
  return this._getURL();
});

Command.prototype.__defineGetter__('args', function () {
  var args = this._getArgs() || {};
  args.timeout = this._args.timeout;
  return args;
});

/**
 * v1: upload, remove, download, getMeta
 */

function UploadCommand(client, args) {
  UploadCommand.super_.call(this, client, args);
}
util.inherits(UploadCommand, Command);

UploadCommand.prototype._getURL = function () {
  var filepath = this._args.filepath;
  var pathname = '/v1/' + this._client.appkey;
  var extname = this._args.extname || path.extname(filepath);
  extname = extname && extname[0] !== '.' ? '.' + extname : extname;
  if (extname) {
    pathname += '?suffix=' + extname + '&simple_name=1';
  }
  return pathname;
};

UploadCommand.prototype._getArgs = function () {
  return {
    type: 'POST',
    dataType: 'json',
    content: this._args.content,
  };
};

function RemoveCommand(client, args) {
  RemoveCommand.super_.call(this, client, args);
  this.ignoreStatus = 404;
}
util.inherits(RemoveCommand, Command);

RemoveCommand.prototype._getURL = function () {
  var url = '/v1/' + this._client.appkey + '/' + this._args.name;
  if (this._args.hide) {
    url += '?hide=' + this._args.hide;
  }
  return url;
};

RemoveCommand.prototype._getArgs = function () {
  return {
    type: 'DELETE',
    headers: { 'Content-Length': 0 }
  };
};

function DownloadCommand(client, args) {
  DownloadCommand.super_.call(this, client, args);
}
util.inherits(DownloadCommand, Command);

DownloadCommand.prototype._getURL = function () {
  var url = '/v1/' + this._client.appkey + '/' + this._args.name;
  var offset = parseInt(this._args.offset, 10);
  var size = parseInt(this._args.size, 10);
  if (!isNaN(offset)) {
    url += '?offset=' + offset;
    if (size > 0) {
      url += '&size=' + size;
    }
  }
  return url;
};

DownloadCommand.prototype._getArgs = function () {
  return {
    type: 'GET',
    writeStream: this._args.writeStream
  };
};

function GetMetaCommand(client, args) {
  GetMetaCommand.super_.call(this, client, args);
}
util.inherits(GetMetaCommand, Command);

GetMetaCommand.prototype._getURL = function () {
  var url = '/v1/' + this._client.appkey + '/metadata/' + this._args.name;
  var type = parseInt(this._args.type, 10);
  if (!isNaN(type)) {
    url += '?type=' + type;
  }
  return url;
};

GetMetaCommand.prototype._getArgs = function () {
  return {
    type: 'GET',
  };
};

/**
 * v2: createFile, uploadFile, removeFile, downloadFile
 */

function CreateFileCommand(client, args) {
  CreateFileCommand.super_.call(this, client, args);
  this.ignoreStatus = 409; // 409 Conflict, dir exists
}
util.inherits(CreateFileCommand, Command);

CreateFileCommand.prototype._getURL = function () {
  var url = CreateFileCommand.super_.prototype._getURL.call(this);
  return url + '?recursive=1';
};

CreateFileCommand.prototype._getArgs = function () {
  return {
    type: 'POST',
    headers: { 'Content-Length': 0 }
  };
};

function UploadFileCommand(client, args) {
  UploadFileCommand.super_.call(this, client, args);
}
util.inherits(UploadFileCommand, Command);

UploadFileCommand.prototype._getURL = function () {
  var url = UploadFileCommand.super_.prototype._getURL.call(this);
  if (typeof this._args.offset === 'number') {
    url += '?offset=' + this._args.offset + '&size=' + this._args.content.length;
  }
  return url;
};

UploadFileCommand.prototype._getArgs = function () {
  return {
    type: 'PUT',
    content: this._args.content,
  };
};

function DownloadFileCommand(client, args) {
  DownloadFileCommand.super_.call(this, client, args);
}
util.inherits(DownloadFileCommand, Command);

DownloadFileCommand.prototype._getArgs = function () {
  var data = {};
  if (this._args.offset) {
    data.offset = this._args.offset;
  }
  if (this._args.size) {
    data.size = this._args.size;
  }
  return {
    type: 'GET',
    data: data,
    writeStream: this._args.writeStream
  };
};

function RemoveFileCommand(client, args) {
  RemoveFileCommand.super_.call(this, client, args);
  this.ignoreStatus = 404;
}
util.inherits(RemoveFileCommand, Command);

RemoveFileCommand.prototype._getArgs = function () {
  return {
    type: 'DELETE',
    headers: { 'Content-Length': 0 }
  };
};

function GetFileMetaCommand(client, args) {
  GetFileMetaCommand.super_.call(this, client, args);
}
util.inherits(GetFileMetaCommand, Command);

GetFileMetaCommand.prototype._getURL = function () {
  var url = '/v2/' + this._client.appkey + '/metadata/' + this._client.appid + '/' +
    this._args.uid + '/file/' + this._args.filename;
  return url;
};

GetFileMetaCommand.prototype._getArgs = function () {
  return {
    type: 'GET',
  };
};

// v1
exports.UploadCommand = UploadCommand;
exports.RemoveCommand = RemoveCommand;
exports.DownloadCommand = DownloadCommand;
exports.GetMetaCommand = GetMetaCommand;

// v2
exports.CreateFileCommand = CreateFileCommand;
exports.UploadFileCommand = UploadFileCommand;
exports.DownloadFileCommand = DownloadFileCommand;
exports.RemoveFileCommand = RemoveFileCommand;
exports.GetFileMetaCommand = GetFileMetaCommand;


demo/buyerstory.js
/*!
 * tfs - demo/buyerstory.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var path = require('path');
var tfs = require('../');

var client = tfs.createClient({
  rootServer: '$domain:$port',
  appkey: 'appkey',
  imageServers: [
    'img01.daily.taobaocdn.net',
    'img02.daily.taobaocdn.net'
  ]
});

client.on('servers', function (servers) {
  console.log(servers);
});

client.on('refreshError', function (err) {
  throw err;
});

var logopath = path.join(path.dirname(__dirname), 'logo.png');

// client.uploadFile(logopath, '320', 'tfs.png', function (err, info) {
//   console.log('uploadFile info: %j', info);
//   client.getMeta(info.name, function (err, meta) {
//     console.log(err)
//     console.log('uploadFile meta: %j', meta);
//   });
// });

client.upload(logopath, function (err, info) {
  console.log('upload info: %j', info);
  client.getMeta(info.name, function (err, meta) {
    // console.log(err)
    console.log('upload meta: %j', meta);
  });
});


lib/client.js
/*!
 * tfs - lib/client.js
 * Copyright(c) 2012 - 2013 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var debug = require('debug')('tfs');
var path = require('path');
var fs = require('fs');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var urllib = require('urllib');
var eventproxy = require('eventproxy');
var Stream = require('stream').Stream;
var bt = require('buffer-type');
var command = require('./command');

/**
 * Create TFS RESTFul client.
 *
 * @param {Object} options
 *  - {String} appkey
 *  - {String} [appLocation], default is 'tfscom'.
 *  - {Number} [uploadTimeout], upload max timeout, default is 60s.
 *  - {String} [rootServer], 'host:port' format.
 *  - {Array} [imageServers], default is CDN online servers list.
 * @return {Client}
 */
function Client(options) {
  options = options || {};
  this.appLocation = options.appLocation || 'tfscom';
  this.appkey = options.appkey;
  if (!this.appkey) {
    throw new TypeError('missing appkey');
  }

  this.rootServer = options.rootServer;
  var pair = this.rootServer.split(':');
  this.rootServerHost = pair[0];
  this.rootServerPort = pair[1] || 80;
  this.imageServers = options.imageServers || [
    'img1.tbcdn.cn',
    'img2.tbcdn.cn',
    'img3.tbcdn.cn',
    'img4.tbcdn.cn',
  ];
  this.imageServersIndex = 0;
  this.uploadTimeout = options.uploadTimeout || 60000;
  this.servers = [];
  this.serverIndex = 0;
  this.refreshCounter = 0;
  this.queue = [];
  this._init();
}

util.inherits(Client, EventEmitter);

Client.prototype._init = function () {
  var that = this;
  that._initTimer = setInterval(that.refreshServers.bind(that), 1000);

  that.once('servers', function () {
    // 初始化完成
    clearInterval(that._initTimer);
    that._initTimer = null;

    // get appid
    that.getAppid(function (err, appid) {
      if (err) {
        // TODO: user how to handle error?
        return that.emit('error', err);
      }
      that.appid = appid;

      // 处理队列
      for (var i = 0; i < that.queue.length; i++) {
        var task = that.queue[i];
        that._request.apply(that, task);
      }
      that.queue = [];
      that.emit('ready');
    });
  });

  // 十分钟更新一次 servers
  setInterval(that.refreshServers.bind(that), 10 * 60 * 1000);
};

var SERVER_HOST_RE = /^\d+\.\d+\.\d+\.\d+/;

Client.prototype.refreshServers = function () {
  var that = this;
  if (that._freshing) {
    return;
  }
  that._freshing = true;
  var options = {
    host: this.rootServerHost,
    port: this.rootServerPort,
    path: '/tfs.list'
  };
  debug('GET /tfs.list');
  urllib.request(options, { timeout: 10000 }, function (err, data, res) {
    var status = res && res.statusCode;
    debug('GET /tfs.list %s', status);
    that._freshing = false;
    if (err) {
      err.url = options.path;
      return that.emit('refreshError', err);
    }

    if (status !== 200) {
      err = new Error('Http Response ' + res.statusCode);
      err.url = options.path;
      err.data = data && data.toString();
      return that.emit('refreshError', err);
    }

    if (data) {
      data = data.toString().trim();
    }

    var lines = data && data.split('\n') || [];
    if (lines.length < 2) {
      return;
    }

    // 应答内容如下，第一行为客户端下次从Webservice Root Server刷新TfsAgent Server的请求间隔，
    // 客户端发送的请求个数达到此数即应该重新从Webservice Root Server获取最新列表，
    // 其下按行存放了所有TfsAgent Server的地址：

    // 50
    // 10.1.1.1:3900
    // 10.1.1.2:3900
    // 10.1.1.3:3900
    var counter = parseInt(lines[0], 10) || 10;
    var servers = [];
    for (var i = 1; i < lines.length; i++) {
      var server = lines[i];
      if (!SERVER_HOST_RE.test(server)) {
        continue;
      }
      servers.push(server.split(':'));
    }
    if (servers.length === 0) {
      return;
    }

    that.refreshCounter = counter;
    that.serverIndex = 0;
    that.servers = servers;
    that.emit('servers', servers);
  });
};

Client.prototype._request = function (cmd, callback) {
  var that = this;
  if (that.servers.length === 0) {
    // add to queue
    return that.queue.push([ cmd, callback ]);
  }

  var server = that.servers[that.serverIndex++];
  if (that.serverIndex >= that.servers.length) {
    that.serverIndex = 0;
  }

  var pathname = cmd.url;
  var args = cmd.args || {};

  var options = {
    host: server[0],
    port: server[1],
    path: pathname,
  };
  args.timeout = args.timeout || that.uploadTimeout;

  debug('%s %s on %j, timeout %s', args.type, pathname, server, args.timeout);
  urllib.request(options, args, function (err, data, res) {
    var status = res && res.statusCode;
    var headers = res && res.headers;

    if (!err) {
      // success
      if (status === 200 || status === 201 || (cmd.ignoreStatus && status === cmd.ignoreStatus)) {
        debug('%s %j', status, headers);
        return callback(null, data);
      }

      // http non 200 error
      err = new Error('TFS request error, Http status ' + status);
      err.name = 'TFSRequestError';
    }

    if (status === 500 && data) {
      data = data.toString();
      var title = /<title>([^<]+)<\/title>/i.exec(data);
      title = title && title[1] || 'Internal Server Error';
      var serverName = /<td>server:<\/td>[^<]+<td>([^<]+)<\/td>/i.exec(data);
      serverName = serverName && serverName[1] || 'unknow server';
      err = new Error('TFS Server error: ' + title + ' (' + serverName + ')');
      err.name = 'TFSServerError';
    }

    err.status = status;
    err.headers = headers;
    err.url = server.join(':') + pathname;
    if (data) {
      if (Buffer.isBuffer(data)) {
        if (data.length > 1024) {
          data = data.slice(0, 1024);
        }
        data = data.toString();
      }
    }
    err.data = data;
    debug('%s %s: %s %s', args.type, pathname, status, err.message);
    callback(err);
  });

  // 判断是否重新获取 server 列表
  that.refreshCounter--;
  if (that.refreshCounter <= 0) {
    that.refreshServers();
  }
};

Client.prototype.getURL = function (pathname) {
  var imgServer = this.imageServers[this.imageServersIndex++];
  if (this.imageServersIndex >= this.imageServers.length) {
    this.imageServersIndex = 0;
  }
  return 'http://' + imgServer + '/' + pathname;
};

Client.prototype.getAppid = function (callback) {
  var pathname = '/v2/' + this.appkey + '/appid';
  var args = {
    type: 'GET',
    dataType: 'json'
  };
  this._request({ url: pathname, args: args }, function (err, info) {
    if (err) {
      return callback(err);
    }
    var appid = info && info.APP_ID;
    if (!appid) {
      err = new Error('GET ' + pathname + ' return appid is empty');
      err.name = 'TFSRequestError';
      return callback(err);
    }
    debug('appid: %s', appid);
    callback(null, appid);
  });
};

Client.prototype.createCommand = function (name, args) {
  var cmd = new command[name + 'Command'](this, args);
  cmd.__name = name;
  return cmd;
};

function getContent(file, callback) {
  if (typeof file === 'string') {
    return fs.readFile(file, callback);
  }
  if (Buffer.isBuffer(file)) {
    return callback(null, file);
  }
  if (file instanceof Stream) {
    var datas = [];
    file.on('data', datas.push.bind(datas));
    file.once('end', function () {
      callback(null, Buffer.concat(datas));
    });
    return;
  }
}

/**
 * v1
 */

/**
 * Upload a file.
 *
 * @param {String|Buffer} filepath file path or file content buffer.
 * @param {String} [extname]. optional, if filepath is buffer or stream type, use this extname.
 * @param {Function(err, info)} callback
 *  - {Object} info
 *   - {String} name, tfs file name
 *   - {String} url, CDN url
 *   - {Number} size, file size
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.upload = function (filepath, extname, callback, timeout) {
  if (typeof extname === 'function') {
    timeout = callback;
    callback = extname;
    extname = null;
  }
  var that = this;
  getContent(filepath, function (err, buffer) {
    if (err) {
      return callback(err);
    }

    if (Buffer.isBuffer(filepath)) {
      filepath = '_buffer';
      if (!extname) {
        // try to guess content-type and extention
        var r = bt(buffer);
        if (r) {
          extname = r.extension;
        }
      }
    }

    var args = {
      filepath: filepath,
      timeout: timeout,
      content: buffer,
      extname: extname
    };

    var cmd = that.createCommand('Upload', args);

    var size = buffer.length;
    debug('upload %d %s', size, filepath);
    that._request(cmd, function (err, data) {
      if (err) {
        return callback(err);
      }
      var name = data && data.TFS_FILE_NAME;
      if (!name) {
        // error
        err = new Error('TFS upload error');
        err.name = 'TFSUploadError';
        err.data = data;
        return callback(err);
      }

      var info = {
        name: name,
        size: size,
        url: that.getURL(that.appLocation + '/' + name)
      };
      debug('%j', info);
      callback(null, info);
    });
  });
};

Client.prototype.uploadPrivate = function (filename, extname, callback, timeout) {
  if (typeof extname === 'function') {
    timeout = callback;
    callback = extname;
    extname = null;
  }
  if (!extname && typeof filename === 'string') {
    extname = path.extname(filename);
    extname = extname && extname.slice(1);
  }

  extname = extname
    ? 'tfsprivate.' + extname
    : 'tfsprivate';

  this.upload(filename, extname, callback, timeout);
};

/**
 * Remove a file by name.
 *
 * @param {String} name
 * @param {Object} [options]
 *  - {Number|String} hide, 1: hidden, 0: show
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.remove = function (name, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }
  var args = {
    name: name,
    timeout: timeout,
  };
  if (options && typeof options.hide === 'number') {
    args.hide = String(options.hide);
  }

  var cmd = this.createCommand('Remove', args);
  this._request(cmd, function (err, data) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

/**
 * Download a file by name.
 *
 * @param {String} name
 * @param {String|WriteStream} savefile, save file path or writable stream.
 * @param {Object} [options]
 *  - {Number} offset
 *  - {Number} size
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.download = function (name, savefile, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }
  var writeStream = savefile;
  if (typeof savefile === 'string') {
    writeStream = fs.createWriteStream(savefile);
  }
  var args = {
    name: name,
    timeout: timeout,
    writeStream: writeStream,
  };
  if (options) {
    args.offset = options.offset;
    args.size = options.size;
  }
  var cmd = this.createCommand('Download', args);
  this._request(cmd, function (err) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

/**
 * Get file meta.
 * @param {String} name
 * @param {Object} [options]
 *  - {Number} type, 0: normal mode, 1: force mode, get meta info even it was deleted.
 * @param {Function(err, meta)} callback
 * @param {Number} timeout
 */
Client.prototype.getMeta = function (name, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }
  var args = {
    name: name,
    timeout: timeout,
  };
  if (options) {
    args.type = options.type;
  }
  var cmd = this.createCommand('GetMeta', args);
  this._request(cmd, function (err, data) {
    if (err) {
      return callback(err);
    }
    var meta = null;
    try {
      meta = JSON.parse(data);
    } catch (e) {
      err = e;
    }
    callback(err, meta);
  });
};

/**
 * v2 uploadFile, downloadFile, removeFile, getFileMeta
 */

/**
 * Get custom file meta.
 *
 * @param {String} uid user id
 * @param {String} filename custom file name
 * @param {Function(err, meta)} callback
 * @param {Number} timeout
 */
Client.prototype.getFileMeta = function (uid, filename, callback, timeout) {
  var args = {
    uid: uid,
    filename: filename,
    timeout: timeout,
  };
  var cmd = this.createCommand('GetFileMeta', args);
  this._request(cmd, function (err, data) {
    if (err) {
      return callback(err);
    }
    var meta = null;
    try {
      meta = JSON.parse(data);
    } catch (e) {
      err = e;
    }
    callback(err, meta);
  });
};

/**
 * Create a file.
 * @param {String} uid, user id
 * @param {[type]} filename, e.g.: 'foo/bar.jpg', 'foo.png'.
 * @param {Function(err, success)} callback
 * @param {Number} [timeout], request timeout
 */
Client.prototype.createFile = function (uid, filename, callback, timeout) {
  var cmd = this.createCommand('CreateFile', {
    uid: uid,
    filename: filename,
    timeout: timeout
  });
  this._request(cmd, function (err, buffer) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

/**
 * Upload a file with custom filename.
 *
 * @param {String} filepath
 * @param {String} uid, user id
 * @param {String} filename
 * @param {Object} [options]
 *  - {Number} offset
 * @param {Function(err, info)} callback
 * @param {Number} [timeout]
 */
Client.prototype.uploadFile = function (filepath, uid, filename, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }

  var that = this;
  var ep = eventproxy.create('content', 'createFile', function (content, success) {
    var args = {
      uid: uid,
      content: content,
      filename: filename,
      timeout: timeout
    };

    options = options || {};
    options.offset = Number(options.offset);

    if (!isNaN(options.offset)) {
      args.offset = options.offset;
    }

    var cmd = that.createCommand('UploadFile', args);

    that._request(cmd, ep.done(function () {
      // http://image_host:port/L1/appid/userid/file_path
      // http://baike.corp.taobao.com/index.php/CS_RD/tfs/http_server
      var name = 'L1/' + that.appid + '/' + uid + '/' + filename;
      var info = {
        name: name,
        url: that.getURL(name),
        size: content.length
      };
      callback(null, info);
    }));
  });

  ep.fail(callback);

  getContent(filepath, ep.done('content'));
  that.createFile(uid, filename, ep.done('createFile'), timeout);
};

/**
 * Download a file by custom name.
 *
 * @param {String} uid
 * @param {String} filename
 * @param {String|WriteStream} savefile, save file path or writable stream.
 * @param {Object} [options]
 *  - {Number} offset
 *  - {Number} size
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.downloadFile = function (uid, filename, savefile, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }
  var writeStream = savefile;
  if (typeof savefile === 'string') {
    writeStream = fs.createWriteStream(savefile);
  }
  var args = {
    uid: uid,
    filename: filename,
    timeout: timeout,
    writeStream: writeStream,
  };
  if (options) {
    args.offset = options.offset;
    args.size = options.size;
  }
  var cmd = this.createCommand('DownloadFile', args);
  this._request(cmd, function (err) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

/**
 * Remove a custom name file.
 *
 * @param {String} uid
 * @param {String} filename
 * @param {Function(err, success)} callback
 * @param {Number} [timeout]
 */
Client.prototype.removeFile = function (uid, filename, callback, timeout) {
  var cmd = this.createCommand('RemoveFile', {
    uid: uid,
    filename: filename
  });

  this._request(cmd, function (err) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

exports.Client = Client;

exports.createClient = function (options) {
  return new Client(options);
};
